src/pool-cl/proxy/CLProxyHook.sol

```diff
@@ -0,0 +1,399 @@
+        }
+
+        for (uint256 i = 0; i < hooks.length; ++i) {
+            ICLHooks hook = hooks[i];
+            bytes32 hookParameters = bytes32(uint256(hook.getHooksRegistrationBitmap()));
+
+            if (!hookParameters.hasOffsetEnabled(HOOKS_AFTER_INITIALIZE_OFFSET)) {
+                continue;
+            }
+
+            if (hook.afterInitialize(sender, key, sqrtPriceX96, tick, hookData) != ICLHooks.afterInitialize.selector) {
+                revert Hooks.InvalidHookResponse();
+            }
+        }
+
+        return this.afterInitialize.selector;
+    }
+
+    function beforeAddLiquidity(
+        address sender,
+        PoolKey calldata key,
+        ICLPoolManager.ModifyLiquidityParams calldata params,
+        bytes calldata hookData
+    ) external override returns (bytes4) {
+        bytes32 parameters = bytes32(uint256(_hooksRegistrationBitmap));
+        if (!parameters.hasOffsetEnabled(HOOKS_BEFORE_ADD_LIQUIDITY_OFFSET)) {
+            revert HookNotImplemented();
+        }
+
+        for (uint256 i = 0; i < hooks.length; ++i) {
+            ICLHooks hook = hooks[i];
+            bytes32 hookParameters = bytes32(uint256(hook.getHooksRegistrationBitmap()));
+
+            if (!hookParameters.hasOffsetEnabled(HOOKS_BEFORE_ADD_LIQUIDITY_OFFSET)) {
+                continue;
+            }
+
+            if (hook.beforeAddLiquidity(sender, key, params, hookData) != ICLHooks.beforeAddLiquidity.selector) {
+                revert Hooks.InvalidHookResponse();
+            }
+        }
+
+        return this.beforeAddLiquidity.selector;
+    }
+
+    function afterAddLiquidity(
+        address sender,
+        PoolKey calldata key,
+        ICLPoolManager.ModifyLiquidityParams calldata params,
+        BalanceDelta delta,
+        bytes calldata hookData
+    ) external override returns (bytes4, BalanceDelta) {
+        bytes32 parameters = bytes32(uint256(_hooksRegistrationBitmap));
+        if (!parameters.hasOffsetEnabled(HOOKS_AFTER_ADD_LIQUIDITY_OFFSET)) {
+            revert HookNotImplemented();
+        }
+
+        BalanceDelta hookDelta = BalanceDeltaLibrary.ZERO_DELTA;
+        for (uint256 i = 0; i < hooks.length; ++i) {
+            ICLHooks hook = hooks[i];
+            bytes32 hookParameters = bytes32(uint256(hook.getHooksRegistrationBitmap()));
+
+            if (!hookParameters.hasOffsetEnabled(HOOKS_AFTER_ADD_LIQUIDITY_OFFSET)) {
+                continue;
+            }
+
+            (bytes4 s, BalanceDelta d) = hook.afterAddLiquidity(sender, key, params, delta, hookData);
+            if (s != ICLHooks.afterAddLiquidity.selector) {
+                revert Hooks.InvalidHookResponse();
+            }
+
+            if (
+                hookParameters.hasOffsetEnabled(HOOKS_AFTER_ADD_LIQUIDIY_RETURNS_DELTA_OFFSET)
+                    && d != BalanceDeltaLibrary.ZERO_DELTA
+            ) {
+                delta = delta - d;
+                hookDelta = hookDelta + d;
```

Comment:
if there was test cases with hookDelta, you will notice a problem around: `how does those hook take/settle with the vault`